src/box/memtx_engine.cc

```diff
@@ -56,6 +56,7 @@
 #include "read_view.h"
 #include "memtx_tuple_compression.h"
 #include "memtx_space.h"
+#include "box.h"
```

Comment:
Making `memtx_engine.cc` depend on `box.h` doesn't look good. Why not add a new header file for recovery triggers?

Reply:
That's what I did initially but @Gerold103 convinced me not to do so: https://github.com/tarantool/tarantool/pull/7938#discussion_r1024632116

Reply:
Minimizing dependencies reduces compilation time. OTOH `box.h` doesn't depend on much so it's probably fine. Let's ask @alyapunov as the main contributor to the memtx code.

Reply:
I don't like this dependency either. Logically memtx_engine (and other engines) should be a small brick on which box stands. Perhaps it is far for ideal, but we should not make it worse. I don't think that compilation time will be affected, but that line of code directly introduces box and memtx_engine cross dependency.

Reply:
Let's ask @Gerold103 as well, since his words were "Lets better do it without new files."

Reply:
I would agree with @alyapunov and @locker if the triggers wouldn't be in box.cc. But they are. Memtx engine depends on box.cc in both solutions. And that is fine IMO. They are a part of libbox both. Separating memtx engine outside of libbox might be possible, but the triggers are the smallest problem of all then. And our new trigger isn't even the first one like that.

So, in total, both solutions make memtx engine depend on box internals. But the current one (no new headers) is 1) smaller, 2) consistent with `box_on_shutdown_trigger_list`. Thus I still vote for not adding new headers.

However I won't block the patch either way.

Reply:
After a chat with @locker we decided to add a callback for memtx to call when indexes are built.
Now the triggers are local to box, and memtx receives a calback which calls the triggers from box.
<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index b75a82c1d..20ca17896 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -168,6 +168,60 @@ static struct cbus_endpoint tx_prio_endpoint;
 
 RLIST_HEAD(box_on_recovery_state);
 
+/**
+ * Recovery states supported by on_recovery_state triggers.
+ * Are positioned in the order of appearance during initial box.cfg().
+ * The only exception are WAL_RECOVERED and INDEXES_BUILT, which might come in
+ * any order, since the moment when secondary indexes are built depends on
+ * box.cfg.force_recovery and box.cfg.memtx_use_mvcc_engine.
+ */
+enum box_recovery_state {
+	/**
+	 * The node has either recovered the snapshot from the disk, received
+	 * the snapshot from the remote master as part of initial join stage or
+	 * has bootstrapped the cluster.
+	 */
+	RECOVERY_STATE_SNAPSHOT_RECOVERED,
+	/**
+	 * The node has either recovered the local WAL files, received the WALs
+	 * from the remote master as part of final join or has bootstrapped the
+	 * cluster.
+	 */
+	RECOVERY_STATE_WAL_RECOVERED,
+	/**
+	 * The node has built secondary indexes for memtx spaces.
+	 */
+	RECOVERY_STATE_INDEXES_BUILT,
+	/**
+	 * The node has synced with remote peers. IOW, it has transitioned from
+	 * "orphan" to "running".
+	 */
+	RECOVERY_STATE_SYNCED,
+	box_recovery_state_MAX,
+};
+
+static const char *box_recovery_state_strs[box_recovery_state_MAX] = {
+	/* [RECOVERY_STATE_SNAPSHOT_RECOVERED] = */
+	"snapshot_recovered",
+	/* [RECOVERY_STATE_WAL_RECOVERED] = */
+	"wal_recovered",
+	/* [RECOVERY_STATE_INDEXES_BUILT] = */
+	"indexes_built",
+	/* [RECOVERY_STATE_SYNCED] = */
+	"synced",
+};
+
+/** Whether the triggers for "synced" recovery stage have already run. */
+static bool recovery_state_synced_is_reached;
+
+int
+box_run_on_recovery_state(enum box_recovery_state state)
+{
+	assert(state >= 0 && state < box_recovery_state_MAX);
+	return trigger_run(&box_on_recovery_state,
+			   (char *)box_recovery_state_strs[state]);
+}
+
 static void
 builtin_events_init(void);
 
@@ -404,9 +458,6 @@ box_wait_ro(bool ro, double timeout)
 	return 0;
 }
 
-/** Whether the triggers for "synced" recovery stage have already run. */
-static bool recovery_state_synced_is_reached;
-
 void
 box_do_set_orphan(bool orphan)
 {
@@ -3691,6 +3742,12 @@ box_free(void)
 	trigger_destroy(&box_on_recovery_state);
 }
 
+static void
+box_on_indexes_built(void)
+{
+	box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
+}
+
 static void
 engine_init()
 {
@@ -3708,7 +3765,8 @@ engine_init()
 				    cfg_geti("strip_core"),
 				    cfg_geti("slab_alloc_granularity"),
 				    cfg_gets("memtx_allocator"),
-				    cfg_getd("slab_alloc_factor"));
+				    cfg_getd("slab_alloc_factor"),
+				    box_on_indexes_built);
 	engine_register((struct engine *)memtx);
 	box_set_memtx_max_tuple_size();
 
@@ -4460,25 +4518,6 @@ box_cfg(void)
 	}
 }
 
-static const char *box_recovery_state_strs[box_recovery_state_MAX] = {
-	/* [RECOVERY_STATE_SNAPSHOT_RECOVERED] = */
-	"snapshot_recovered",
-	/* [RECOVERY_STATE_WAL_RECOVERED] = */
-	"wal_recovered",
-	/* [RECOVERY_STATE_INDEXES_BUILT] = */
-	"indexes_built",
-	/* [RECOVERY_STATE_SYNCED] = */
-	"synced",
-};
-
-int
-box_run_on_recovery_state(enum box_recovery_state state)
-{
-	assert(state >= 0 && state < box_recovery_state_MAX);
-	return trigger_run(&box_on_recovery_state,
-			   (char *)box_recovery_state_strs[state]);
-}
-
 /**
  * box.coredump() forks to save a core. The entire
  * server forks in box.cfg{} if background=true.
diff --git a/src/box/box.h b/src/box/box.h
index af7f3b31e..33851c8a8 100644
--- a/src/box/box.h
+++ b/src/box/box.h
@@ -75,42 +75,6 @@ extern struct rlist box_on_shutdown_trigger_list;
  */
 extern struct rlist box_on_recovery_state;
 
-/**
- * Recovery states supported by on_recovery_state triggers.
- * Are positioned in the order of appearance during initial box.cfg().
- * The only exception are WAL_RECOVERED and INDEXES_BUILT, which might come in
- * any order, since the moment when secondary indexes are built depends on
- * box.cfg.force_recovery and box.cfg.memtx_use_mvcc_engine.
- */
-enum box_recovery_state {
-	/**
-	 * The node has either recovered the snapshot from the disk, received
-	 * the snapshot from the remote master as part of initial join stage or
-	 * has bootstrapped the cluster.
-	 */
-	RECOVERY_STATE_SNAPSHOT_RECOVERED,
-	/**
-	 * The node has either recovered the local WAL files, received the WALs
-	 * from the remote master as part of final join or has bootstrapped the
-	 * cluster.
-	 */
-	RECOVERY_STATE_WAL_RECOVERED,
-	/**
-	 * The node has built secondary indexes for memtx spaces.
-	 */
-	RECOVERY_STATE_INDEXES_BUILT,
-	/**
-	 * The node has synced with remote peers. IOW, it has transitioned from
-	 * "orphan" to "running".
-	 */
-	RECOVERY_STATE_SYNCED,
-	box_recovery_state_MAX,
-};
-
-/** Run the on_recovery_state_triggers with state \a state. */
-int
-box_run_on_recovery_state(enum box_recovery_state state);
-
 /**
  * Timeout during which the transaction must complete,
  * otherwise it will be rolled back.
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 9a1becaed..52212d64e 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -429,7 +429,7 @@ memtx_engine_begin_final_recovery(struct engine *engine)
 		memtx->state = MEMTX_OK;
 		if (space_foreach(memtx_build_secondary_keys, memtx) != 0)
 			return -1;
-		box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
+		memtx->on_indexes_built_cb();
 	}
 	return 0;
 }
@@ -448,7 +448,7 @@ memtx_engine_begin_hot_standby(struct engine *engine)
 		memtx->state = MEMTX_OK;
 		if (space_foreach(memtx_build_secondary_keys, memtx) != 0)
 			return -1;
-		box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
+		memtx->on_indexes_built_cb();
 	}
 	return 0;
 }
@@ -468,7 +468,7 @@ memtx_engine_end_recovery(struct engine *engine)
 		memtx->state = MEMTX_OK;
 		if (space_foreach(memtx_build_secondary_keys, memtx) != 0)
 			return -1;
-		box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
+		memtx->on_indexes_built_cb();
 	}
 	xdir_collect_inprogress(&memtx->snap_dir);
 
@@ -638,7 +638,7 @@ memtx_engine_bootstrap(struct engine *engine)
 	xlog_cursor_close(&cursor, false);
 	if (rc < 0)
 		return -1;
-	box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
+	memtx->on_indexes_built_cb();
 	return 0;
 }
 
@@ -1263,7 +1263,8 @@ struct memtx_engine *
 memtx_engine_new(const char *snap_dirname, bool force_recovery,
 		 uint64_t tuple_arena_max_size, uint32_t objsize_min,
 		 bool dontdump, unsigned granularity,
-		 const char *allocator, float alloc_factor)
+		 const char *allocator, float alloc_factor,
+		 memtx_on_indexes_built_cb cb)
 {
 	int64_t snap_signature;
 	struct memtx_engine *memtx =
@@ -1370,6 +1371,8 @@ memtx_engine_new(const char *snap_dirname, bool force_recovery,
 	}
 	tuple_format_ref(memtx->func_key_format);
 
+	memtx->on_indexes_built_cb = cb;
+
 	fiber_start(memtx->gc_fiber, memtx);
 	return memtx;
 fail:
diff --git a/src/box/memtx_engine.h b/src/box/memtx_engine.h
index 2fbc74685..8f259c5ca 100644
--- a/src/box/memtx_engine.h
+++ b/src/box/memtx_engine.h
@@ -109,6 +109,9 @@ enum memtx_reserve_extents_num {
  */
 #define MEMTX_ITERATOR_SIZE (168)
 
+typedef void
+(*memtx_on_indexes_built_cb)(void);
+
 struct memtx_engine {
 	struct engine base;
 	/** Engine recovery state, see enum memtx_recovery_state description. */
@@ -126,6 +129,11 @@ struct memtx_engine {
 	 * needed to be able to cancel it on shutdown.
 	 */
 	struct cord *replica_join_cord;
+	/**
+	 * A callback run once memtx engine builds secondary indexes for the
+	 * data.
+	 */
+	memtx_on_indexes_built_cb on_indexes_built_cb;
 	/** Common quota for tuples and indexes. */
 	struct quota quota;
 	/**
@@ -209,7 +217,8 @@ struct memtx_engine *
 memtx_engine_new(const char *snap_dirname, bool force_recovery,
 		 uint64_t tuple_arena_max_size, uint32_t objsize_min,
 		 bool dontdump, unsigned granularity,
-		 const char *allocator, float alloc_factor);
+		 const char *allocator, float alloc_factor,
+		 memtx_on_indexes_built_cb cb);
 
 int
 memtx_engine_recover_snapshot(struct memtx_engine *memtx,
@@ -341,13 +350,13 @@ static inline struct memtx_engine *
 memtx_engine_new_xc(const char *snap_dirname, bool force_recovery,
 		    uint64_t tuple_arena_max_size, uint32_t objsize_min,
 		    bool dontdump, unsigned granularity,
-		    const char *allocator, float alloc_factor)
+		    const char *allocator, float alloc_factor,
+		    memtx_on_indexes_built_cb cb)
 {
 	struct memtx_engine *memtx;
 	memtx = memtx_engine_new(snap_dirname, force_recovery,
-				 tuple_arena_max_size,
-				 objsize_min, dontdump,
-				 granularity, allocator, alloc_factor);
+				 tuple_arena_max_size, objsize_min, dontdump,
+				 granularity, allocator, alloc_factor, cb);
 	if (memtx == NULL)
 		diag_raise();
 	return memtx; 
```

</details>